<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta http-equiv="X-UA-Compatible" content="ie=edge">
  <title>取消请求</title>
</head>
<body>
  <button onclick="getProducts1()">获取商品列表1</button><br>
  <button onclick="getProducts2()">获取商品列表2</button><br>
  <button onclick="cancelReq()">取消请求</button><br>

  <script src="https://cdn.bootcss.com/axios/0.19.0/axios.js"></script>
  <script>
    let cancel;  // 用于保存取消请求的函数
    function getProducts1() {
      axios({
        url: 'http://localhost:4000/products1',
        cancelToken: new axios.CancelToken((c) => { // c是用于取消当前请求的函数
          // 保存取消函数, 用于之后可能需要取消当前请求
          cancel = c;
        })
      }).then(
        response => {
          cancel = null;
          console.log('请求1成功了', response.data);
        },
        error => {
          cancel = null;
          console.log('请求1失败了', error.message, error);
        }
      )
    }

    function getProducts2() {
      axios({
        url: 'http://localhost:4000/products2'
      }).then(
        response => {
          cancel = null;
          console.log('请求2成功了', response.data);
        },
        error => {
          cancel = null;
          console.log('请求2失败了', error.message);
        }
      )
    }

    function cancelReq() {
      // alert('取消请求')
      // 执行取消请求的函数
      if (typeof cancel === 'function') {
        cancel('强制取消请求');
      } else {
        console.log('没有可取消的请求');
      }
    }

    // 取消请求的情况
    // 1. 可能都没有执行请求，就点了取消(这个时候肯定无效，所以需要判断一下)
    // 2. 请求完成后再点击取消，这个时候也没有(因此完成请求了，也需要处理一下)


    // 一旦取消了请求
    // 整个流程就进入了失败的流程
    // 然后进入失败的回调函数
    // 然后这个时候 error 对象的类型是 Cancel 类型 (并不是平常的 Error 类型，因此也没有 stack 属性)
    

  </script>
</body>
</html>